File: examples/web-python/pages/inventory_page.py
from playwright.sync_api import Locator, Page


class InventoryPage:
    def __init__(self, page: Page) -> None:
        self.page = page
        self.title: Locator = page.get_by_text("Products", exact=True)
        self.items: Locator = page.locator(".inventory_item")
        self.cart_badge: Locator = page.locator(".shopping_cart_badge")

    def goto(self) -> None:
        self.page.goto("/inventory.html")

    def item(self, name: str) -> Locator:
        """The card for one product, so actions stay scoped to it."""
        return self.items.filter(has_text=name)

    def add_to_cart(self, name: str) -> None:
        self.item(name).get_by_role("button", name="Add to cart").click()

File: examples/web-ts/pages/InventoryPage.ts
import { type Locator, type Page } from '@playwright/test';

export class InventoryPage {
  readonly title: Locator;
  readonly items: Locator;
  readonly cartBadge: Locator;

  constructor(private readonly page: Page) {
    this.title = page.getByText('Products', { exact: true });
    this.items = page.locator('.inventory_item');
    this.cartBadge = page.locator('.shopping_cart_badge');
  }

  async goto(): Promise<void> {
    await this.page.goto('/inventory.html');
  }

  /** The card for one product, so actions stay scoped to it. */
  item(name: string): Locator {
    return this.items.filter({ hasText: name });
  }

  async addToCart(name: string): Promise<void> {
    await this.item(name).getByRole('button', { name: 'Add to cart' }).click();
  }
}
